src/main/scala/cromwell/engine/SymbolTable.scala

```diff
@@ -0,0 +1,58 @@
+package cromwell.engine
+
+import cromwell.binding.types.WdlType
+import cromwell.binding.{FullyQualifiedName, WdlBinding, WdlValue}
+
+import scala.collection.mutable
+
+case class SymbolTableEntry(scope: String, name: String, iteration: Option[Int], wdlValue: Option[WdlValue], wdlType: WdlType, input: Boolean) {
+
+  wdlValue match {
+    case Some (value) => wdlType.checkCompatible(value.value)
+    case None => // Nothing is always okay for a value.
+  }
+
+  def isInput: Boolean = input
+
+  def isOutput: Boolean = !isInput
+
+  /**
+   * Create a copy of the receiver with `wdlValue` set to `newValue`.
+   */
+  def copyWithUpdatedValue(newValue: WdlValue): SymbolTableEntry =
+    SymbolTableEntry(scope, name, iteration, Option(newValue), wdlType, input)
+}
+
+class SymbolTable(binding: WdlBinding, inputs: Map[FullyQualifiedName, WdlValue]) {
+  val table = mutable.Set[SymbolTableEntry]()
+
+  binding.workflow.calls.foreach { call =>
+    call.task.inputs.foreach { case (inputName, wdlType) =>
+      table += new SymbolTableEntry(call.fullyQualifiedName, inputName, iteration = None, wdlValue = None, wdlType, input = true)
+    }
+    call.task.outputs.foreach { output =>
+      table += new SymbolTableEntry(call.fullyQualifiedName, output.name, iteration = None, wdlValue = None, output.wdlType, input = false)
+    }
+  }
+
+  private def assignSymbolTableEntry(fullyQualifiedName: FullyQualifiedName, value: WdlValue, iteration: Option[Int], input: Boolean): Unit = {
+    val lastIndex = fullyQualifiedName.lastIndexOf(".")
+    val (scope, name) = (fullyQualifiedName.substring(0, lastIndex), fullyQualifiedName.substring(lastIndex + 1))
+
+    val maybeOldEntry = table.find { entry =>
```

Comment:
Yeah, that's why it's no bueno.

I suspect that if we're having these sorts of conversations that 'case' no longer should proceed 'class'. But it might be worth taking a brief step back and seeing if an alternative approach would make any sense (probably not).

If you think of an instance of a case class as a single immutable instance of a single type (which is what it really is), the fact that we're trying to do this and that is really breaking that model. The only other thought which comes to mind (and adds complexity thus the 'probably not') would be to have a case class around the stuff which is actually case class-y and then a wrapper class which contains that + the extra bits. 

I actually kinda like that, you've effectively made a primary key for that object. But it's an extra layer of indirection.